graph/src/blockchain/block_stream.rs

```diff
@@ -410,6 +418,18 @@ pub trait SubstreamsMapper<C: Blockchain>: Send + Sync {
                 )))
             }
 
+            Some(SubstreamsMessage::Progress(progress)) => {
+                if last_progress.elapsed() > Duration::from_secs(30) {
+                    info!(
+                        &logger,
+                        "Received progress update";
+                        "progress" => format!("{:?}", progress),
```

Comment:
The `progress` message in full is a bit verbose and hard to work with when printed using `Debug` trait.

I see different possibilities. Quickest one would be to print as JSON, we would need to tweak a bit the generated code so that `ModulesProgress` can be serialized to JSON. See commit https://github.com/graphprotocol/graph-node/commit/d66e348bd4b87fcdc636f36123496641fa9cacc2 for details.

My preferred approach would be that in `INFO`, we present a condensed "view" of the `ModulesProgress`, I would see:

```
Substreams backend `graph_out` last block is <last_seen_block>, Stages: [#S up to B (<first(reversed modules)>, +M), ...], Jobs: [#J on Stage S @ <start_block> | +<processed_blocks>|<end_block - start_block> elapsed <formatted duration>, ...]
```

The rendering would look like this:

```
Substreams backend `graph_out` last block is 18332001, Stages: [#5 (graph_out), #4 up to 18332000 (store_derived_factory_tvl), #3 up to 18332000 (store_max_windows, +3), #2 up to 16332000 (store_eth_prices), #1 up to 18332000 (store_ticks_liquidities, +10), #0 up to 18332000 (store_pool_count, +3)], Jobs: [#0 on Stage 4 @ 18332000 +517/1000 elapsed 3m 53s], [#1 on Stage 3 @ 18332000 +747/1000 elapsed 3m 53s], [#2 on Stage 2 @ 18332000 +877/1000 elapsed 3m 53s], [#3 on Stage 1 @ 18332000 +898/1000 elapsed 3m 53s], [#4 on Stage 0 @ 18332000 +894/1000 elapsed 3m 53s], [#5 on Stage 0 @ 18333000 +0/2 elapsed 3m 53s]
```

In `DEBUG` it could be a more expanded view (for example all modules instead of +N) and in `TRACE` could be the JSON view.

If you guys decide to make it, I can define more specifically how and where to get the info I have in my formatted view.

Reply:
I don't see a reason why it wouldn't be done that way - each level with increased details. Maybe the info could be even more condensed...

Reply:
@maoueh can you tell me where to get the info for the more concise log levels, please?

Reply:
Using format and assuming Rust struct progress message of type [ModulesProgress](https://github.com/graphprotocol/graph-node/pull/4935/files#diff-c67185a22602db5b79b5ccf8449efd8f79530a243ae472729665b54c753c3dcdR174):

> `Substreams backend `graph_out` last block is <last_seen_block>, Stages: [#S up to B (<first(reversed modules)>, +M), ...], Jobs: [#J on Stage S @ <start_block> | +<processed_blocks>|<end_block - start_block> elapsed <formatted duration>, ...]`

- `<last_seen_block>` Would be kept within `substreams_block_stream` struct directly, would be initialized to start block and then would be updated as the stream receives block (would update on `BlockScopedData` and `BlockUndoSignal` messages).
- `Stages` array would be printed in reverse order as found in `ModulesProgress.stages` field of the Rust struct.
- `#S` represents the **original** index in the `stages` struct (I say original so that index of reversed array is not used)
- `up to B` this should be displayed if the `Stage.completed_ranges.len()` is > 0 where the `B` value is (pseudo-code here) `Stage.completed_ranges.map(x.end_block).min()`.
- `(<first(reversed modules)>, +M)`, my definition here is a bit complicate, so feel free to change. I would render the `Stage.modules.reverse()` dynamically based on the module count. If module count is 1, render as `(<module_name>)` otherwise renders as `(first(reverse_modules), +{reverse_modules.len()-1})`. 
- `Jobs` would be the in order `ModulesProgress.running_jobs`
- `#J` would be the index of the Job within `running_jobs`
- `on Stage S` would refer to `Job.stage`.
- `<start_block>` would be `Job.start_block`
- `+<processed_blocks>` would be `Job.processed_blocks`
- `<end_block - start_block>` is actually the block count to be executed by the job, so `Job.end_block - Job.start_block`.
- `<formatted duration>` would be humanized value of `Job.duration_ms`.

Hopefully this is clear enough, happy to chat over Slack if further details is needed on what I had in mind.

We could think of a summary for `INFO`, what I propose above in `DEBUB` and JSON for `TRACE`. I leave you guys the decision on those.